Documentation/upgrading.md

```diff
@@ -1,173 +1,149 @@
+# Upgrading self-hosted Kubernetes
 
-# Upgrading Self-hosted Kubernetes
+"Self-hosted" Kubernetes clusters run the apiserver, scheduler, controller-manager, kube-dns, kube-proxy, and flannel or calico as pods, like ordinary applications (except with taint tolerations). This allows upgrades to be performed in-place using (mostly) `kubectl` as an alternative to re-provisioning.
 
-Self-hosted Kubernetes clusters schedule Kubernetes components such as the apiserver, kubelet, scheduler, and controller-manager as pods like other applications (except with node selectors). This allows Kubernetes level operations to be performed to upgrade clusters in place, rather than by re-provisioning.
+Let's upgrade a Kubernetes v1.6.6 cluster to v1.6.7 as an example.
 
-Let's upgrade a self-hosted Kubernetes v1.4.1 cluster to v1.4.3 as an example.
+## Status
 
-## Inspect
-
-Show the control plane daemonsets and deployments which will need to be updated.
-
-    $ kubectl get daemonsets -n=kube-system
-    NAME             DESIRED   CURRENT   NODE-SELECTOR                    AGE
-    kube-apiserver   1         1         node-role.kubernetes.io/master   5m
-    kube-proxy       3         3         <none>                           5m
-    kubelet          3         3         <none>                           5m
-
-    $ kubectl get deployments -n=kube-system
-    NAME                      DESIRED   CURRENT   UP-TO-DATE   AVAILABLE   AGE
-    kube-controller-manager   1         1         1            1           5m
-    kube-dns-v20              1         1         1            1           5m
-    kube-scheduler            1         1         1            1           5m
-
-Check the current Kubernetes version.
-
-    $ kubectl version
-    Client Version: version.Info{Major:"1", Minor:"4", GitVersion:"v1.4.0", GitCommit:"a16c0a7f71a6f93c7e0f222d961f4675cd97a46b", GitTreeState:"clean", BuildDate:"2016-09-26T18:16:57Z", GoVersion:"go1.6.3", Compiler:"gc", Platform:"linux/amd64"}
-    Server Version: version.Info{Major:"1", Minor:"4", GitVersion:"v1.4.1+coreos.0", GitCommit:"b7a02f46b972c5211e5c04fdb1d5b86ac16c00eb", GitTreeState:"clean", BuildDate:"2016-10-11T20:13:55Z", GoVersion:"go1.6.3", Compiler:"gc", Platform:"linux/amd64"}
-
-In this case, Kubernetes is `v1.4.1+coreos.0` and our goal is to upgrade to `v1.4.3+coreos.0`. First, update the control plane pods. Then the kubelets and proxies on all nodes.
-
-## Control Plane
-
-### kube-apiserver
+bootkube does not provide in-place patch diffs or migrations between releases. In-place upgrades are inherently more dangerous than a blue/green cluster deployments. Evaluate whether an in-place upgrade is appropriate for your Kubernetes cluster and risk tolerance before beginning. Failover important workloads to another cluster in advance.
```

Comment:
Maybe something like:

> While the process of upgrading a self-hosted cluster can be relatively straight-forward, there may be underlying functionality changes that could affect your cluster. 

> In most cases patch version upgrades (e.g. `v1.6.6` to `v1.6.7`) are safe to apply in-place. 

>For minor version upgrades (e.g. `v1.6.7` to `v1.7.0`) it is important to evaluate if an in-place upgrade is appropriate for your Kubernetes cluster. A recommendation would be to first test the in-place upgrade on a development cluster that uses the same configurations.

Reply:
The optimistic lead in sounds fine to me.

I'd still hope to recommend people "failover important workloads to another cluster in advance" or "weight the dangers against blue/green deployments". When managing clusters for teams and sufficiently important apps, there's no amount of "I tested in-place upgrades on a development cluster" that make this safe. At a minimum, there should be a 2nd cluster ready. Otherwise we're going to get panic'd questions like "site is down, how to recover now!".

Reply:
Updated to something vaguely in between. What do you think?